test/runtests.jl

```diff
@@ -1047,6 +1047,99 @@ function test_model_filename()
     return
 end
 
+function test_model_nlp_boolean()
+    model = MOI.Utilities.UniversalFallback(MOI.Utilities.Model{Int}())
+    x = MOI.add_variables(model, 2)
+    MOI.add_constraint.(model, x, MOI.ZeroOne())
+    nlp = MOI.Nonlinear.Model()
+    a, b = x
+    MOI.Nonlinear.add_constraint(nlp, :(($a || $b)), MOI.EqualTo(1.0))
+    MOI.Nonlinear.add_constraint(nlp, :(($a && $b)), MOI.EqualTo(0.0))
+    backend = MOI.Nonlinear.ExprGraphOnly()
+    evaluator = MOI.Nonlinear.Evaluator(nlp, backend, x)
+    MOI.set(model, MOI.NLPBlock(), MOI.NLPBlockData(evaluator))
+    solver = MiniZinc.Optimizer{Int}(MiniZinc.Chuffed())
+    MOI.set(solver, MOI.RawOptimizerAttribute("model_filename"), "test.mzn")
+    index_map, _ = MOI.optimize!(solver, model)
+    @test MOI.get(solver, MOI.TerminationStatus()) === MOI.OPTIMAL
+    @test MOI.get(solver, MOI.ResultCount()) >= 1
+    y = [index_map[v] for v in x]
+    sol = round.(Bool, MOI.get(solver, MOI.VariablePrimal(), y))
+    @test (sol[1] || sol[2]) == true
+    @test (sol[1] && sol[2]) == false
+    @test read("test.mzn", String) ==
+          "var bool: x1;\nvar bool: x2;\nconstraint (x1 \\/ x2) == true;\nconstraint (x1 /\\ x2) == false;\nsolve satisfy;\n"
+    rm("test.mzn")
+    return
+end
+
+function test_model_nlp_boolean_nested()
+    model = MOI.Utilities.UniversalFallback(MOI.Utilities.Model{Int}())
+    x = MOI.add_variables(model, 2)
+    MOI.add_constraint.(model, x, MOI.ZeroOne())
+    y = MOI.add_variable(model)
+    MOI.add_constraint(model, y, MOI.Integer())
+    MOI.add_constraint(model, y, MOI.Interval(0, 10))
+    nlp = MOI.Nonlinear.Model()
+    a, b = x
+    # a || (b && (y < 5))
+    MOI.Nonlinear.add_constraint(
+        nlp,
+        :(($a || ($b && ($y < 5)))),
```

Comment:
@chriscoey nested expressions like this work, for both binary and integer variables.

Reply:
Very nice!! 🤩

Reply:
I'll merge so that you can experiment with it more easily. It'll have some rough edges (assertion errors) when things go wrong, but we can fix those in time if there is interest.

Reply:
awesome yes i plan to experiment with it in the next few days. thanks @odow!